core/src/main/java/software/amazon/awssdk/core/auth/ContainerCredentialsProvider.java

```diff
@@ -0,0 +1,153 @@
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.annotations.SdkTestInternalApi;
+import software.amazon.awssdk.core.AwsSystemSetting;
+import software.amazon.awssdk.core.SdkClientException;
+import software.amazon.awssdk.core.internal.CredentialsEndpointProvider;
+import software.amazon.awssdk.core.retry.internal.CredentialsEndpointRetryPolicy;
+import software.amazon.awssdk.utils.StringUtils;
+
+/**
+ * {@link AwsCredentialsProvider} implementation that loads credentials from the Amazon Elastic Container Service.
+ *
+ * <p>The URI path is retrieved from the environment variable "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI" in the container's
+ * environment. If the environment variable is not set, this credentials provider will throw an exception.</p>
+ */
+public final class ContainerCredentialsProvider extends HttpCredentialsProvider {
+    private static final Logger log = LoggerFactory.getLogger(ContainerCredentialsProvider.class);
+    private final CredentialsEndpointProvider credentialsEndpointProvider;
+
+    /**
+     * @see #builder()
+     */
+    private ContainerCredentialsProvider(Builder builder) {
+        super(builder);
+        this.credentialsEndpointProvider = builder.credentialsEndpointProvider;
+    }
+
+    /**
+     * Create a builder for creating a {@link ContainerCredentialsProvider}.
+     */
+    static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    protected CredentialsEndpointProvider getCredentialsEndpointProvider() {
+        return credentialsEndpointProvider;
+    }
+
+    @Override
+    public String toString() {
+        return getClass().getSimpleName();
+    }
+
+    static final class ContainerCredentialsEndpointProvider implements CredentialsEndpointProvider {
+        private static final Set<String> ALLOWED_HOSTS = unmodifiableSet(new HashSet<>(Arrays.asList("localhost", "127.0.0.1")));
+
+        @Override
+        public URI endpoint() throws IOException {
+
+            if (!AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_RELATIVE_URI.getStringValue().isPresent() &&
+                !AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.getStringValue().isPresent()) {
+                throw new SdkClientException(
+                    String.format("Cannot fetch credentials from container - neither %s or %s environment variables are set.",
+                                  AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_FULL_URI.environmentVariable(),
+                                  AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_RELATIVE_URI.environmentVariable()));
+            }
+
+            try {
+                return AwsSystemSetting.AWS_CONTAINER_CREDENTIALS_RELATIVE_URI.getStringValue().map(relativeUri -> {
+                    return URI.create(AwsSystemSetting.AWS_CONTAINER_SERVICE_ENDPOINT.getStringValueOrThrow() + relativeUri);
```

Comment:
can we remove `return` and the curly braces here?

Reply:
👍